File: 0872-leaf-similar-trees/0872-leaf-similar-trees.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def leafSimilar(self, root1: Optional[TreeNode], root2: Optional[TreeNode]) -> bool:
        def collect_leaves(root):
            if not root: return []
            return collect_leaves(root.left) + ([root.val] if not root.left and not root.right else []) + collect_leaves(root.right)
        
        return collect_leaves(root1) == collect_leaves(root2)

File: 0872-leaf-similar-trees/0872-leaf-similar-trees.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void find(TreeNode* root, vector<int> &v)
    {
        if(root == NULL)
            return;
        
        if(root->left == NULL && root->right == NULL)
        {
            v.push_back(root->val);
            return;
        }
        
        find(root->left,v);
        find(root->right,v);
    }
    
    bool leafSimilar(TreeNode* root1, TreeNode* root2) {
        vector<int> v1,v2;
        
        find(root1,v1);
        find(root2,v2);
        
        if(v1 == v2)
            return true;
        
        return false;
    }
};